### PR TITLE
make clk id configurable

### PR DESCRIPTION
**Pull request by YangKeao**
Signed-off-by: Yang Keao <keao.yang@yahoo.com>

### What problem does this PR solve?

It makes the mask of affected `clk_id` configurable.

### What is changed and how does it work?

Add a new variable in `fake_clock_gettime` function.

### Check List 

Tests 

 - Unit test
 - Manual test

Code changes

 - Has Go code change

### CODE REVIEW

**Comment by cwen0** [comment]
@ethercflow PTAL

File: cmd/watchmaker/main_linux.go [hunk 1]
```diff
@@ -29,16 +32,18 @@ import (
...
+	nsecDelta     int64
+	printVersion  bool
+	clockIdsSlice string
 )
 
 func initFlag() {
 	flag.IntVar(&pid, "pid", 0, "pid of target program")
 	flag.Int64Var(&secDelta, "sec_delta", 0, "delta time of sec field")
 	flag.Int64Var(&nsecDelta, "nsec_delta", 0, "delta time of nsec field")
+	flag.StringVar(&clockIdsSlice, "clk_ids", "CLOCK_REALTIME", "affected clock ids")
```
Comment (ethercflow) [inline cmd/watchmaker/main_linux.go]: The fourth parameter may describe the split is ',' ?

File: pkg/time/time_linux.go [hunk 2]
```diff
@@ -137,14 +145,20 @@ func ModifyTime(pid int, deltaSec int64, deltaNsec int64) error {
 	}
 	fakeAddr := fakeEntry.StartAddress
 
-	// 115 is the index of TV_SEC_DELTA in fakeImage
-	err = program.WriteUint64ToAddr(fakeAddr+115, uint64(deltaSec))
+	// 139 is the index of CLOCK_IDS_MASK in fakeImage
```
Comment (ethercflow) [inline pkg/time/time_linux.go]: The order of the index does not match the 79 comment rows?

File: cmd/watchmaker/main_linux.go [hunk 3]
```diff
@@ -61,7 +66,15 @@ func main() {
 	ptrace.RegisterLogger(log.WithName("ptrace"))
 	time.RegisterLogger(log.WithName("time"))
 
-	err = time.ModifyTime(pid, secDelta, nsecDelta)
+	clkIds := strings.Split(clockIdsSlice, ",")
+	mask, err := utils.EncodeClkIds(clkIds)
+	if err != nil {
+		log.Error(err, "error while converting clock ids to mask")
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply cmd/watchmaker/main_linux.go]: `log` from `controller-runtime` will bring in lots of flags about kubernetes, which is disappointing. And it uses `zapr` too, so literally using this `log` doesn't bring in any new dependency.

File: api/v1alpha1/timechaos_types.go [hunk 4]
```diff
@@ -53,6 +53,11 @@ type TimeChaosSpec struct {
 	// TimeOffset defines the delta time of injected program
 	TimeOffset TimeOffset `json:"timeOffset"`
 
+	// ClockIds defines all affected clock id
+	// All available options are ["CLOCK_REALTIME","CLOCK_MONOTONIC","CLOCK_PROCESS_CPUTIME_ID","CLOCK_THREAD_CPUTIME_ID","CLOCK_MONOTONIC_RAW","CLOCK_REALTIME_COARSE","CLOCK_MONOTONIC_COARSE","CLOCK_BOOTTIME","CLOCK_REALTIME_ALARM","CLOCK_BOOTTIME_ALARM"]
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply api/v1alpha1/timechaos_types.go]: Fixed

File: cmd/watchmaker/main_linux.go [hunk 5]
```diff
@@ -61,7 +66,15 @@ func main() {
...
+		return
+	}
+	log.Info("get clock ids mask", "mask", mask)
+
+	err = time.ModifyTime(pid, secDelta, nsecDelta, mask)
 
 	if err != nil {
 		fmt.Printf("error while modifying time, pid: %d, sec_delta: %d, nsec_delta: %d\n Error: %s", pid, secDelta, nsecDelta, err.Error())
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply cmd/watchmaker/main_linux.go]: Fixed

File: doc/time_chaos.md [hunk 6]
```diff
@@ -41,5 +41,4 @@ Description:
 
 * Time modification will only be injected into the main process of container
 * Time chaos has no effect on pure syscall `clock_gettime`
-* Only `CLOCK_REALTIME` will be modified. This is intended because modification on other clock such as `CLOCK_REALTIME_COARSE` (which go program uses to `time.Sleep`) will cause serious and deterministic problem. We should make it configurable in the future.
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply doc/time_chaos.md]: Yes! I added description in former part

File: pkg/utils/time.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+import "fmt"
+
+// EncodeClkIds will convert array of clk ids into a mask
+func EncodeClkIds(clkIds []string) (uint64, error) {
+	mask := uint64(0)
+
+	for _, id := range clkIds {
+		// refer to `uapi/linux/time.h`
+		if id == "CLOCK_REALTIME" {
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/utils/time.go]: Fixed

File: pkg/time/time_linux.go (hunk 2 above)
Comment (YangKeao) [reply pkg/time/time_linux.go]: I should modify 77-79 comments. They should follow the order of being used in program.

File: cmd/watchmaker/main_linux.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cwen0) [reply cmd/watchmaker/main_linux.go]: got

**Comment by cwen0** [comment]
@ethercflow PTAL

**Comment by cwen0** [comment]
@YangKeao Fix conflicts

File: helm/chaos-mesh/values.yaml [hunk 8]
```diff
@@ -19,7 +19,7 @@ controllerManager:
   replicaCount: 1
 
   image: pingcap/chaos-mesh:latest
-  imagePullPolicy: Always
+  imagePullPolicy: Never
```
Comment (yeya24) [inline helm/chaos-mesh/values.yaml]: Can I ask why this change to `Never`?
